File: src/main/kotlin/com/bookd/domain/service/parser/TxtParser.kt
package com.bookd.domain.service.parser

import com.bookd.domain.model.ContentElement
import com.bookd.domain.model.TextSpan
import com.bookd.domain.service.TxtParseRuleService
import org.slf4j.LoggerFactory
import java.io.File

class TxtParser(
    private val txtParseRuleService: TxtParseRuleService
) {
    
    private val logger = LoggerFactory.getLogger(TxtParser::class.java)
    
    data class TxtStructure(
        val chapters: List<ChapterInfo>,
        val fullText: String
    )
    
    data class ChapterInfo(
        val index: Int,
        val title: String?,
        val startPos: Int,
        val endPos: Int,
        val level: Int = 0
    )
    
    /**
     * 获取启用的章节解析规则
     */
    private suspend fun getChapterPatterns(): List<Regex> {
        return try {
            val rules = txtParseRuleService.getEnabledRules()
            if (rules.isEmpty()) {
                logger.warn("No enabled TXT parse rules found, using fallback patterns")
                getFallbackPatterns()
            } else {
                rules.mapNotNull { rule ->
                    try {
                        rule.rule.toRegex()
                    } catch (e: Exception) {
                        logger.error("Invalid regex pattern for rule '${rule.name}': ${rule.rule}", e)
                        null
                    }
                }
            }
        } catch (e: Exception) {
            logger.error("Failed to load TXT parse rules, using fallback patterns", e)
            getFallbackPatterns()
        }
    }
    
    /**
     * 备用模式（如果数据库规则加载失败）
     */
    private fun getFallbackPatterns(): List<Regex> = listOf(
        // 英文: Chapter 1, Chapter One, Ch.1, Ch 1
        """^(Chapter|CHAPTER|Ch\.|Ch)\s*([0-9]+|One|Two|Three|Four|Five|Six|Seven|Eight|Nine|Ten).*$""".toRegex(),
        
        // 中文数字: 第一章, 第1章, 第一节, 第1节, 第一回, 第1回
        """^第([零一二三四五六七八九十百千万0-9]+)[章节回].*$""".toRegex(),
        
        // 简化: 一、二、三、
        """^([零一二三四五六七八九十百千万]+)[、．.].*$""".toRegex(),
        
        // 阿拉伯数字: 1., 1、, 01.
        """^([0-9]{1,3})[、．.].*$""".toRegex(),
        
        // 卷/部/篇
        """^第([零一二三四五六七八九十百千万0-9]+)[卷部篇].*$""".toRegex()
    )
    
    /**
     * 解析 TXT 文件结构
     */
    suspend fun parseStructure(file: File): TxtStructure? {
        try {
            val fullText = TxtFileDecoder.read(file)
            val chapters = detectChapters(fullText)
            
            logger.info("Detected ${chapters.size} chapters in TXT file")
            return TxtStructure(chapters, fullText)
        } catch (e: Exception) {
            logger.error("Failed to parse TXT structure: ${file.name}", e)
            return null
        }
    }
    
    /**
     * 检测章节分割
     */
    private suspend fun detectChapters(text: String): List<ChapterInfo> {
        val chapterPatterns = getChapterPatterns()
        val chapters = mutableListOf<ChapterInfo>()
        val lines = splitLinesWithOffsets(text)
        
        var currentChapterStart = 0
        var chapterIndex = 0
        var lastChapterLine = 0
        lines.forEachIndexed { lineIndex, line ->
            val trimmedLine = line.text.trim()
            
            // 跳过空行
            if (trimmedLine.isNotEmpty()) {
                // 检查是否匹配章节标题
                chapterPatterns.forEach { pattern ->
                    if (pattern.matches(trimmedLine)) {
                        // 如果不是第一章，保存前一章
                        if (chapterIndex > 0) {
                            chapters.add(
                                ChapterInfo(
                                    index = chapterIndex - 1,
                                    title = lines[lastChapterLine].text.trim(),
                                    startPos = currentChapterStart,
                                    endPos = line.startOffset
                                )
                            )
                        }
                        
                        // 开始新章节
                        currentChapterStart = line.startOffset
                        lastChapterLine = lineIndex
                        chapterIndex++
                        
                        logger.debug("Found chapter $chapterIndex: $trimmedLine")
                        return@forEachIndexed
                    }
                }
            }
            
        }
        
        // 添加最后一章或整本书作为一章
        if (chapterIndex > 0) {
            chapters.add(
                ChapterInfo(
                    index = chapterIndex - 1,
                    title = lines[lastChapterLine].text.trim(),
                    startPos = currentChapterStart,
                    endPos = text.length
                )
            )
        } else {
            // 没有检测到章节，整本书作为一章
            chapters.add(
                ChapterInfo(
                    index = 0,
                    title = "全文",
                    startPos = 0,
                    endPos = text.length
                )
            )
        }
        
        return chapters
    }

    /**
     * 保留原始换行符所占的字符位置，避免 CRLF/CR 文本的章节区间逐行漂移。
     */
    private fun splitLinesWithOffsets(text: String): List<TextLine> {
        val lines = mutableListOf<TextLine>()
        var lineStart = 0
        var index = 0
        while (index < text.length) {
            val char = text[index]
            if (char == '\r' || char == '\n') {
                lines.add(TextLine(text.substring(lineStart, index), lineStart))
                index += if (char == '\r' && index + 1 < text.length && text[index + 1] == '\n') 2 else 1
                lineStart = index
            } else {
                index++
            }
        }
        lines.add(TextLine(text.substring(lineStart), lineStart))
        return lines
    }

    private data class TextLine(val text: String, val startOffset: Int)
    
    /**
     * 提取章节内容
     */
    fun extractChapterContent(fullText: String, chapterInfo: ChapterInfo): List<ContentElement> {
        val chapterText = fullText.substring(chapterInfo.startPos, chapterInfo.endPos)
        val anchorGenerator = ContentAnchorGenerator(
            sourceKind = "txt",
            chapterIdentity = chapterInfo.index.toString()
        )
        
        // 统一换行符为 \n，然后按行分割
        val normalizedText = chapterText.replace("\r\n", "\n").replace("\r", "\n")
        val lines = normalizedText.split("\n")
        
        val elements = mutableListOf<ContentElement>()
        var isFirstNonEmpty = true
        
        lines.forEach { line ->
            val trimmedLine = line.trim()
            
            if (trimmedLine.isEmpty()) {
                // 空行跳过
                return@forEach
            }
            
            // 第一个非空行如果是标题，添加为 Heading
            if (isFirstNonEmpty && chapterInfo.title != null && trimmedLine.startsWith(chapterInfo.title.trim())) {
                elements.add(
                    ContentElement.Heading(
                        level = 1,
                        text = trimmedLine,
                        anchorId = anchorGenerator.generatedAnchor("heading", trimmedLine)
                    )
                )
                isFirstNonEmpty = false
            } else {
                // 每行作为一个独立段落
                elements.add(
                    ContentElement.Paragraph(
                        spans = listOf(TextSpan(trimmedLine)),
                        anchorId = anchorGenerator.generatedAnchor("paragraph", trimmedLine)
                    )
                )
                isFirstNonEmpty = false
            }
        }
        
        return elements
    }
    
    /**
     * 计算章节字数
     */
    fun countWords(text: String): Int {
        // 简单计数：中文按字符数，英文按单词数
        val chineseCount = text.count { it.code in 0x4E00..0x9FFF }
        val englishWords = text.split(Regex("\\s+"))
            .count { it.matches(Regex("[a-zA-Z]+")) }
        
        return chineseCount + englishWords
    }
}

File: ebook-parser/app/txt_parser.py
"""TXT file parser for chapter detection and content extraction."""

import re
import logging
from typing import List, Tuple, Optional

from app.models import (
    TxtParseRule,
    TxtChapterInfo,
    TxtStructureResponse,
    TxtContentResponse,
    ContentElement,
    TextSpan
)

logger = logging.getLogger(__name__)


class TxtParser:
    """Parser for TXT files with customizable chapter detection rules."""
    
    # Fallback patterns (used when no custom rules provided)
    FALLBACK_PATTERNS = [
        # English: Chapter 1, Chapter One, Ch.1, Ch 1
        r'^(Chapter|CHAPTER|Ch\.|Ch)\s*([0-9]+|One|Two|Three|Four|Five|Six|Seven|Eight|Nine|Ten).*$',
        
        # Chinese numbers: 第一章, 第1章, 第一节, 第1节, 第一回, 第1回
        r'^第([零一二三四五六七八九十百千万0-9]+)[章节回].*$',
        
        # Simplified: 一、二、三、
        r'^([零一二三四五六七八九十百千万]+)[、．.].*$',
        
        # Arabic numbers: 1., 1、, 01.
        r'^([0-9]{1,3})[、．.].*$',
        
        # Volume/Part/Book
        r'^第([零一二三四五六七八九十百千万0-9]+)[卷部篇].*$'
    ]
    
    def parse_structure(self, file_path: str, rules: List[TxtParseRule] = None) -> TxtStructureResponse:
        """Parse TXT file structure to detect chapters.
        
        Args:
            file_path: Absolute path to TXT file
            rules: Custom parse rules (if empty, use fallback patterns)
            
        Returns:
            TxtStructureResponse with detected chapters
        """
        try:
            # Read file with multiple encoding attempts
            full_text = self._read_file(file_path)
            if full_text is None:
                return TxtStructureResponse(
                    chapters=[],
                    total_chapters=0,
                    full_text_length=0,
                    success=False,
                    error="Failed to read file with supported encodings"
                )
            
            # Get chapter patterns
            patterns = self._get_patterns(rules)
            
            # Detect chapters
            chapters = self._detect_chapters(full_text, patterns)
            
            logger.info(f"Detected {len(chapters)} chapters in TXT file: {file_path}")
            
            return TxtStructureResponse(
                chapters=chapters,
                total_chapters=len(chapters),
                full_text_length=len(full_text),
                success=True
            )
            
        except Exception as e:
            logger.error(f"Failed to parse TXT structure: {e}", exc_info=True)
            return TxtStructureResponse(
                chapters=[],
                total_chapters=0,
                full_text_length=0,
                success=False,
                error=str(e)
            )
    
    def extract_chapter_content(
        self,
        file_path: str,
        start_pos: int,
        end_pos: int,
        chapter_title: Optional[str] = None
    ) -> TxtContentResponse:
        """Extract chapter content from TXT file.
        
        Args:
            file_path: Absolute path to TXT file
            start_pos: Start position (character index)
            end_pos: End position (character index)
            chapter_title: Chapter title (optional)
            
        Returns:
            TxtContentResponse with content elements
        """
        try:
            # Read file
            full_text = self._read_file(file_path)
            if full_text is None:
                return TxtContentResponse(
                    elements=[],
                    word_count=0,
                    success=False,
                    error="Failed to read file"
                )
            
            # Extract chapter text
            chapter_text = full_text[start_pos:end_pos]
            
            # Normalize line breaks
            normalized_text = chapter_text.replace('\r\n', '\n').replace('\r', '\n')
            lines = normalized_text.split('\n')
            
            # Convert to content elements
            elements = []
            is_first_non_empty = True
            
            for line in lines:
                trimmed_line = line.strip()
                
                if not trimmed_line:
                    # Skip empty lines
                    continue
                
                # First non-empty line: check if it's the chapter title
                if is_first_non_empty and chapter_title and trimmed_line.startswith(chapter_title.strip()):
                    elements.append(ContentElement(
                        type="heading",
                        level=1,
                        text=trimmed_line
                    ))
                    is_first_non_empty = False
                else:
                    # Each line is a paragraph
                    elements.append(ContentElement(
                        type="paragraph",
                        spans=[TextSpan(text=trimmed_line)]
                    ))
                    is_first_non_empty = False
            
            # Count words
            word_count = self._count_words(chapter_text)
            
            return TxtContentResponse(
                elements=elements,
                word_count=word_count,
                success=True
            )
            
        except Exception as e:
            logger.error(f"Failed to extract chapter content: {e}", exc_info=True)
            return TxtContentResponse(
                elements=[],
                word_count=0,
                success=False,
                error=str(e)
            )
    
    def _read_file(self, file_path: str) -> Optional[str]:
        """Read file with multiple encoding attempts.
        
        Args:
            file_path: Absolute path to file
            
        Returns:
            File content as string, or None if failed
        """
        encodings = ['utf-8', 'gbk', 'gb2312', 'big5', 'utf-16', 'iso-8859-1']
        
        for encoding in encodings:
            try:
                with open(file_path, 'r', encoding=encoding) as f:
                    content = f.read()
                logger.debug(f"Successfully read file with encoding: {encoding}")
                return content
            except (UnicodeDecodeError, LookupError):
                continue
            except Exception as e:
                logger.error(f"Failed to read file with encoding {encoding}: {e}")
                continue
        
        logger.error(f"Failed to read file with all supported encodings: {file_path}")
        return None
    
    def _get_patterns(self, rules: Optional[List[TxtParseRule]]) -> List[re.Pattern]:
        """Get regex patterns from rules or fallback patterns.
        
        Args:
            rules: Custom parse rules
            
        Returns:
            List of compiled regex patterns
        """
        if rules and len(rules) > 0:
            # Sort by priority
            sorted_rules = sorted(rules, key=lambda r: r.priority)
            
            patterns = []
            for rule in sorted_rules:
                if not rule.enabled:
                    continue
                
                try:
                    pattern = re.compile(rule.rule)
                    patterns.append(pattern)
                except re.error as e:
                    logger.error(f"Invalid regex pattern for rule '{rule.name}': {rule.rule}, error: {e}")
                    continue
            
            if patterns:
                logger.info(f"Using {len(patterns)} custom parse rules")
                return patterns
            else:
                logger.warning("No valid custom rules, falling back to default patterns")
        
        # Use fallback patterns
        logger.info("Using fallback chapter detection patterns")
        return [re.compile(p) for p in self.FALLBACK_PATTERNS]
    
    def _detect_chapters(self, text: str, patterns: List[re.Pattern]) -> List[TxtChapterInfo]:
        """Detect chapters in text using patterns.
        
        Args:
            text: Full text content
            patterns: Compiled regex patterns
            
        Returns:
            List of detected chapters
        """
        chapters = []
        lines = text.split('\n')
        
        current_chapter_start = 0
        chapter_index = 0
        last_chapter_line = 0
        current_position = 0
        
        for line_index, line in enumerate(lines):
            trimmed_line = line.strip()
            line_length = len(line) + 1  # +1 for newline
            
            # Skip empty lines
            if trimmed_line:
                # Check if line matches any chapter pattern
                for pattern in patterns:
                    if pattern.match(trimmed_line):
                        # Save previous chapter (if not first)
                        if chapter_index > 0:
                            chapters.append(TxtChapterInfo(
                                index=chapter_index - 1,
                                title=lines[last_chapter_line].strip(),
                                start_pos=current_chapter_start,
                                end_pos=current_position,
                                level=0
                            ))
                        
                        # Start new chapter
                        current_chapter_start = current_position
                        last_chapter_line = line_index
                        chapter_index += 1
                        
                        logger.debug(f"Found chapter {chapter_index}: {trimmed_line}")
                        break  # Don't check other patterns for this line
            
            current_position += line_length
        
        # Add last chapter or entire text as one chapter
        if chapter_index > 0:
            chapters.append(TxtChapterInfo(
                index=chapter_index - 1,
                title=lines[last_chapter_line].strip(),
                start_pos=current_chapter_start,
                end_pos=len(text),
                level=0
            ))
        else:
            # No chapters detected, use entire text as one chapter
            chapters.append(TxtChapterInfo(
                index=0,
                title="全文",
                start_pos=0,
                end_pos=len(text),
                level=0
            ))
        
        return chapters
    
    def _count_words(self, text: str) -> int:
        """Count words in text.
        
        For Chinese text: counts characters
        For English text: counts words (space-separated)
        
        Args:
            text: Text content
            
        Returns:
            Word count
        """
        # Count Chinese characters (CJK Unified Ideographs)
        chinese_count = len(re.findall(r'[\u4e00-\u9fff]', text))
        
        # Count English words
        english_words = re.findall(r'[a-zA-Z]+', text)
        english_count = len(english_words)
        
        return chinese_count + english_count
